### DISCUSSION TITLE
How are you using laravel-messenger?

### DISCUSSION THREAD

**Discussion by cmgmyr** [cmgmyr/laravel-messenger#366] [Show and tell]
Please leave a comment explaining how you are using this package. How many users are you supporting? What type of features have you had to implement to extend the base functionality of this package?

_*If you have any issues, please open up a new ticket. Please keep this thread for feedback only. Thanks!_

**Comment by mrgwilliam** [comment]
Implementing it for a Unit Manager internal messaging system. We aim to support at most 150+ users. Currently only using base functionality. I'll update I add anything more. Thanks for this awesome package!

**Comment by nuukcillo** [comment]
We are using it for the same that mrgwilliam, but we are also implementating a notification system with other parts of the site. We have around 500 users in the site but the messenger is still in beta.

**Comment by Pandahisham** [comment]
we are implementing it in an in-school network , however , having seen laravel 5.1 shouldbroadcast , we are thinking about using socket.io to build a realtime messaging platform , have you given any thought about this ? there's a package called 'confer' that already does this but with pusher

**Comment by mydnic** [comment]
We are using it as a facebook-like private message system in a community website.
Thanks for this package !

**Comment by cmgmyr** [comment]
Thank you all for the great feedback so far!

@Pandahisham in the new version I'm planning on using events, so you'll be able to hook into those pretty easily. You'll still have to handle the front end processing on your own. I don't think it makes sense (at least at this time) to do anything with Pusher, or other services, inside this package

**Comment by Pandahisham** [comment]
yap ! we will handle the frontend ourselfves , just give us a well documented package with hints to how it may work with say pusher or socket.io , leave the rest to us

**Comment by IJack** [comment]
Thanks for the nice package, works fine. 

But I got an error when I make a migration:refresh after a time: Class "AddSoftdeletesToThreadTable" not found. The migration file with it class is in the migration folder. Don't know what I'm doing wrong

**Comment by antonkomarev** [comment]
@Pandahisham just check this repo. It's pretty good boilerplate to make this package work with Pusher https://github.com/cmgmyr/laravel-messenger-pusher-demo

**Comment by Pandahisham** [comment]
@a-komarev and @cmgmyr thanks !! i may have to re-purpose that to socket.io though

**Comment by talha08** [comment]
How can I run this in my project , I have no idea... .. I followed all the instructions but its not working  :( I am using laravel 4.2 version ...anyone here to help me please!!

**Comment by cmgmyr** [comment]
@talha08 if you have a specific thing you need help with, please create a new issue ticket with more information. This isn't the correct issue for help. With Laravel 4, you should be using the `1.x` version. I hope that helps.

**Comment by FrankDupree** [comment]
@cmgmyr  Using it as a personal messaging system with the phphub china community forum.
you could check it out here. [Phphub++](https://github.com/FrankDupree/phphub)
![](https://cloud.githubusercontent.com/assets/8538930/12395169/0e63c0f8-bdf8-11e5-8c25-f27a7abf6494.jpg)

**Comment by Gummibeer** [comment]
In a Game-Webinterface for an Arma3 Altis Life Server.

AS additional features I've added Markdown and a PHP side Twemoji Class that replaces some codes with the SVGs for the twemojis.

For me this is a very good package cause it just comes with the needed things and is easy customizable.

**Comment by bgies** [comment]
I'm not using it yet, but I'm planning on using it to support messaging between the website admins and the users... it seems very well suited for that

**Comment by sger** [comment]
Great package just created a demo app with Laravel 5.2

https://github.com/sger/laravel-messenger-demo

**Comment by cmgmyr** [comment]
Thanks for the demo @sger! I've been toying with the idea of making a live working demo that people can log in to, but just haven't had the time recently. Definitely on my to-do list though.

**Comment by 3traker** [comment]
> Great package just created a demo app with Laravel 5.2
> 
> https://github.com/sger/laravel-messenger-demo

404 :(

**Comment by SAIBA0011** [comment]
I'm using it for a friends messaging system :) Planning on using this for a open forum type as well.

**Comment by sger** [comment]
@3traker @SAIBA0011 download the project here https://www.dropbox.com/s/4dc7cwa5x5ihc3p/laravel-messenger-demo.zip?dl=0

**Comment by 3traker** [comment]
Thanks!

**Comment by xuan9230** [comment]
Hi, in my website there are two different kinds of users(multi auth), let's say "cop" and "thief" instead of one kind of "user". Can I still use this package?
Thanks!

**Comment by antonkomarev** [comment]
@xuan9230 of course. You can extend Thread model or even write your own to fit your requirements. This package provides a structure which you can adapt for your needs.

**Comment by xuan9230** [comment]
@a-komarev So I've noticed "user" appears a lot in several files (Thread, MessengerServiceProvider, Participant...) Should I change every single one of them? That seems to be such a huge workload...

**Comment by cmgmyr** [comment]
@xuan9230 nope, all you have to do is publish the config file and update the reference to the "user" model you want [here](https://github.com/cmgmyr/laravel-messenger/blob/master/src/config/config.php). All of the references to your model will get mapped automatically. Also, if you have any questions/issues with the package, please open a new ticket so we can reserve this one for feedback related comments

**Comment by xuan9230** [comment]
@cmgmyr thanks man! I will open a new ticket once I figured out the entire process:)

**Comment by groundbreaker08** [comment]
@cmgmyr implementing it on our private network! :D Thank you for this package.

**Comment by Moussa112** [comment]
@cmgmyr Realy nice package you have here! Thanks a lot! I would like to use it in my project as some sort of private direct messaging service betweens users. But I can't seem to find a way to make it "private". I used to code provided in the controller to show only the threads you are participating in so that's cool but you can still go to the "messages" you are not participating in by changing the url like so for example: "messages/1". Is there a way provided in your package to make this secure? Thanks in advance!

**Comment by cmgmyr** [comment]
@Moussa112 thanks for using the package! There isn't anything built directly into this package for private one-to-one messaging, I left that up to the developer since there are multiple ways that you can implement it. Please take a look at #29 for some helpful info. I'm going to be writing up some additional information about how to do this soon.

**Comment by cihadturhan** [comment]
We implemented an internal messaging module in our [private course website](http://www.derstakvimi.com) thanks to your package. Also, we made some modifications to send bulk messages to everyone.

![image](https://cloud.githubusercontent.com/assets/3739536/16356623/b3f0b6ea-3ae7-11e6-88ce-5a9d4e0ec76d.png)

**Comment by cmgmyr** [comment]
Thanks for sharing @cihadturhan, the site looks great! I really like the page layout too. How did you handle the bulk notifications? Just sending out one-to-one messages to everyone from an admin user?

**Comment by Naveen95** [comment]
Any idea how much simultaneous connections it can manage ? planning to implement it on a project where tons of chats may take place.

**Comment by cmgmyr** [comment]
@Naveen95 that's really up to your server infrastructure. Laravel can handle many concurrent connections. If you want more of a real-time chat, you should probably look into other technologies instead of PHP. This really depends on your project and how you want to implement it.

**Comment by naumanahmed19** [comment]
any working demo ?

**Comment by cmgmyr** [comment]
@naumanahmed19 unfortunately, nothing "official" at the moment, but if you look at my repos you'll see a number of tests in different versions of the framework that you can pull down and play with. They all come with migrations and seeds to get you up and running.

**Comment by lankaen** [comment]
@cmgmyr  any idea how to developing realtime application using laravel 5.2 using sockets [ I tried vinelab/minion (built using crossbar.io) but it's not working correctly in my laravel environment ]

**Comment by rajjanorkar** [comment]
My goal was to implement one to one messaging system using this package. I must say this is really awesome package, with little changes to example controller and created my own views and it's live in few hours.

We implemented it in [JustBeFriend.com](http://JustBeFriend.com)

Thank you very much everyone who contributed to this repo, and keep up the good work.

**Comment by yansusanto** [comment]
I'm looking for a system where it's Admin -> many Users / one User and user -> Admin only. No messaging between users. 

Does this package fit our requirement?

Thanks ;)

PS: Any real live demo would be good though

**Comment by cmgmyr** [comment]
@yanseo Yes, you can use this package for that. You'd have to build the rules around who can contact who though, but the base messaging functionality would still be the same.

**Comment by watchwood** [comment]
How easy would it be to send system generated messages to users with this package?

**Comment by Promixius** [comment]
Such functionallity is build into laravel 
https://laravel.com/docs/5.3/notifications

**Comment by cmgmyr** [comment]
@watchwood - like @Promixius stated, notifications would be better for this. You _could_ use this package to accomplish something similar, but it's certainly not ideal (IMO)

**Comment by h-azad** [comment]
Does it compatible with Laravel 5.3 ??

**Comment by cmgmyr** [comment]
@sidor555 [yes](https://github.com/cmgmyr/test-messenger-laravel53)

**Comment by h-azad** [comment]
Hi... When I'm trying to migrate database it's showing. I already have a messages table in my database. I think this maybe a problem -->  Cannot redeclare class CreateMessagesTable ..... So now what to do ??

**Comment by cmgmyr** [comment]
@sidor555 can you please open a new issue with more information?

**Comment by reliq** [comment]
We're using your package [here](https://github.com/reliqarts/mardin) and at [studentroomsja.com](http://studentroomsja.com).

Thank you for an awesome package man!

:beers: cheers!

**Comment by cmgmyr** [comment]
Very cool @reliq, thanks for sharing!

**Comment by asimshazad** [comment]
Sorry @reliq ,i can't find this package on studentroomsja.com :(

**Comment by reliq** [comment]
Hey @asimshazad, we're basically it for the member chat feature on that site (http://stuentroomsja.com). The feature isn't actually open to the public. The chat is currently only available to people with 'room requests'.

**Comment by Rostom22** [comment]
Hi, can i add attached-file fonctionnality to this package ? thanks.

**Comment by antonkomarev** [comment]
@Rostom22 For questions its better to make issues. This thread used to share projects where this package is used. Answer to your question: #162

**Comment by aarontanzk** [comment]
Hi anyone has one-to-one messenging feature example project?

**Comment by DavisAre** [comment]
Thank you for creating this package. Although I have created similar PM systems in plain-php, I didn't want to re-invent the wheel this time. Took some hours of modifying and this is the result. It's an implementation for my bachelor's thesis project. ;)
![My example](https://i.imgur.com/i9StkXh.png)

**Comment by herryswastika** [comment]
I use Laravel. I combine the laravel-messenger with email notification with Mailgun and Amazon SES. Working very very good. User can also send message directly to Admin without knowing who the Admin is (with admin role). I also use Summernote rich text editor so laravel-messenger message body could contain image, picture, video link, etc. I have planning to create chat directly from Post model, but still considering it given that comment section is provided. User can also opt to 'GetMeOut' if he/she does not longer want to be included in the thread message (although I assume his/her boss will get angry because the App is intended to be used in a company)
![thread_lq](https://user-images.githubusercontent.com/33825938/52801840-57174c00-30b1-11e9-9272-7d38bbd7969d.png)
![thread_admin](https://user-images.githubusercontent.com/33825938/52802141-e9b7eb00-30b1-11e9-86e3-f25622d410ed.png)
![thread_notifemail](https://user-images.githubusercontent.com/33825938/52804450-cfccd700-30b6-11e9-87f0-1edc6e135e22.png)
![thread_unsub](https://user-images.githubusercontent.com/33825938/52804453-d3605e00-30b6-11e9-9852-8b057eae52bb.png)

**Comment by cmgmyr** [comment]
so cool, thanks for sharing!

**Comment by mustafa-online** [comment]
can I use this package with two separate models? "User" and "Manager" ??

**Comment by simonj** [comment]
@Mu9tafa-Online I want that too 👍

**Comment by yudi-glive** [comment]
We're using this package to provide instant messages to our users. Glive connects artists, agencies and contractors. They can meet new contacts and keep in touch with messages. We're developing more advanced messaging features like contract attachments, event planning and more.

This package allows the customization of the message system to our needs. Thank you for this excellent package.

https://glivesol.com
![messenger_demo_2](https://user-images.githubusercontent.com/54516686/66790598-cd7d7f80-eec6-11e9-9020-45bac5a81518.png)

**Comment by alkaou** [comment]
Please helps me here, i have a serious problem, use Laravel 5.8 i really need your helps

https://stackoverflow.com/questions/59249412/get-the-users-who-are-sent-themselves-of-the-messages-with-the-last-messages-and

**Comment by cmgmyr** [comment]
@alkaou this thread is not for asking for help. Please add a [new, detailed, issue](https://github.com/cmgmyr/laravel-messenger/issues/new) in the repo or wait for someone to respond on stack overflow.

**Comment by RTippin** [comment]
Actually started using this package along with an expanded version of this package (https://github.com/lexxyungcarter/laravel-5-messenger) as a base in April 2018, but ended up extracting out your SRC files, using the models as a base reference, and building back up from scratch. We needed to be able to message between any models we choose, so switched to using morphs on everything. Ended up building so much on top, I am starting to extract it out into a standalone package I too would like to release as an entire suite! But your base helped us get the jump start we needed, so thanks! I added support for group invite links like discord, video calling/group calling, bobble heads (read indicators), images, documents, links, youtube support. Privacy and friends list are also integrated

![71425952-e5187780-2670-11ea-8390-583970b6ba82](https://user-images.githubusercontent.com/3917769/71426168-d67f8f80-2673-11ea-9b4f-289b72b490fa.png)

![71425992-5821ee00-2671-11ea-94a5-488215856885](https://user-images.githubusercontent.com/3917769/71426169-d8e1e980-2673-11ea-93a6-87802aa92762.png)

**Comment by dendoy** [comment]
thanks for an awesome package....
![Dashboard_BinaKarir](https://user-images.githubusercontent.com/37052991/73901721-39488580-48c6-11ea-8e48-f12c3e0aa0bd.png)

**Comment by debuGhy** [comment]
> thanks for an awesome package....
> ![Dashboard_BinaKarir](https://user-images.githubusercontent.com/37052991/73901721-39488580-48c6-11ea-8e48-f12c3e0aa0bd.png)

Hello I know its bit late to ask a question. But I would like to know how to show the conversations in timeline mode like you are doing?
I am using $thread->messages to display all the messages between two users. But I want to display the message from the sender like you are doing in blue and the receiver of the message in white. I want to implement the conversation view like you are doing. How to achieve it? Kindly help.

**Comment by pionas** [comment]
Hello. I am rebuilding my website and looked a package for messages between users. Your package looks nice but i added two changes.

Firstly. In Thread model i added new scope:
```
    /**
     * Returns threads that the user is associated with and sorts start from unreads.
     *
     * @param \Illuminate\Database\Eloquent\Builder $query
     * @param int $userId
     *
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function scopeForUserOrderByNotReadMessages(Builder $query, $userId)
    {
        $participantTable = Models::table('participants');
        $threadsTable = Models::table('threads');

        $orderBy = 'IF(`' . $participantTable . '`.last_read IS NULL, 2, IF (`' . $threadsTable . '`.updated_at>`' . $this->getConnection()->raw($this->getConnection()->getTablePrefix() . $participantTable) . '`.last_read, 2, 0)) DESC';
        return $query->join($participantTable, $this->getQualifiedKeyName(), '=', $participantTable . '.thread_id')
                ->where($participantTable . '.user_id', $userId)
                ->whereNull($participantTable . '.deleted_at')
                ->orderByRaw($orderBy)
                ->select($threadsTable . '.*');
    }
```

Secondly. I modified getParticipantFromUser method.
```
    public function getParticipantFromUser($userId)
    {
        if ($this->relationLoaded('participants')) {
            $participant = $this->participants->where('user_id', $userId)->first();
            if (is_null($participant)) {
                throw new ModelNotFoundException();
            }
            return $participant;
        }
        return $this->participants()->where('user_id', $userId)->firstOrFail();
    }
```

And i used this:
```
Thread::forUserOrderByNotReadMessages($userID)->with(['participants' => function ($query) use ($userID) {
        $query->where('user_id', $userID);
    }])->latest('updated_at')->paginate(50);
```

These changes reduce the number of queries and the page loads faster :)

**Comment by BrunoBasstos** [comment]
I'm studying laravel and I've decided to rebuild an application using it.
One of the new features will be users chat and your package was just what I needed.
Made some adjustments to star and archive messages and thats it!
Thanks a lot for the great work!
Here is how it looks...

![chat-demo](https://user-images.githubusercontent.com/5402439/88055316-2a345780-cb35-11ea-8f20-664ab2f2f87e.png)

**Comment by ghost** [comment]
Hello every bady
![88055316-2a345780-cb35-11ea-8f20-664ab2f2f87e](https://user-images.githubusercontent.com/61189903/93053382-26c6d380-f681-11ea-86d5-31db3ec74aad.png)
Mana natija

**Comment by prog-24** [comment]
Using this project to power messaging on our platform.
![simulator_screenshot_7272839B-FDA6-4E30-B920-FA228BBB9ADF](https://user-images.githubusercontent.com/2653516/114248369-680d3080-998f-11eb-9216-69fd47226659.png)

It works great.

**Reply by hetmann** [reply]
@prog-24 native app or cross-platform?
